gtest/src/balance.rs

```diff
@@ -0,0 +1,188 @@
+// This file is part of Gear.
+
+// Copyright (C) 2024 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Balance management.
+
+use std::collections::HashMap;
+
+use crate::{manager::Actors, GAS_MULTIPLIER};
+use gear_common::{Gas, GasMultiplier, ProgramId};
+use gear_core::message::Value;
+
+#[derive(Debug, Clone, Default)]
+pub struct Balance {
+    total: u128,
+    // Primary used for ED locking
+    locked: u128,
+}
+
+impl Balance {
+    pub fn new(total: u128) -> Self {
+        Self { total, locked: 0 }
+    }
+
+    #[track_caller]
+    pub fn set_lock(&mut self, value: Value) {
+        if self.available() < value {
+            unreachable!(
+                "Trying to lock more then available balance, total: {}, lock: {}",
+                self.available(),
+                value
+            );
+        }
+        self.locked = value;
+    }
+
+    pub fn empty() -> Self {
+        Self {
+            total: 0,
+            locked: 0,
+        }
+    }
+
+    #[track_caller]
+    pub fn available(&self) -> u128 {
+        self.total - self.locked
+    }
+
+    pub fn total(&self) -> u128 {
+        self.total
+    }
+
+    #[track_caller]
+    pub fn decrease(&mut self, value: u128, keep_alive: bool) {
+        if keep_alive {
+            if self.available() < value {
+                unreachable!(
+                    "Not enough balance to decrease, available: {}, value: {}",
+                    self.available(),
+                    value
+                );
+            }
+        } else if self.total < value {
+            unreachable!(
+                "Not enough balance to decrease, total: {}, value: {}",
+                self.total, value
+            );
+        }
+
+        self.total -= value;
+        if !keep_alive && self.total < self.locked {
+            self.locked = self.total;
+        }
+    }
+
+    pub fn increase(&mut self, value: u128) {
```

Comment:
please check that `self.total + value > ED`

Reply:
and panic otherwise
also add to `System` method `fn transfer(&mut self, from: impl Into<ProgramIdWrapper>, to: impl Into<ProgramIdWrapper>, value: Value, keep_alive: bool)`

discussed

Reply:
Added check,

```rust
    pub fn increase(&mut self, value: Value) {
        if self.total + value < EXISTENTIAL_DEPOSIT {
            panic!(
                "After increasing, the balance is lower than the existential deposit, total: {}, value: {value}",
                self.total,
            );
        }

        self.total += value;
    }
```

but it seems this check is **incorrect**, 
it will always panic on message consumed in `program::tests::test_handle_exit_with_zero_balance` test.

Reply:
Discussed, gonna fix test